Src/GoogleApis.Core/Apis/Util/PageStreamer.cs

```diff
@@ -0,0 +1,137 @@
+﻿// Copyright 2015 Google Inc. All Rights Reserved.
+// Licensed under the Apache License Version 2.0.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Apis.Util
+{
+    /// <summary>
+    /// A page streamer is a helper to provide both synchronous and asynchronous page streaming
+    /// of a listable or queryable resource.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// The expected usage pattern is to create a single paginator for a resource collection,
+    /// and then use the instance methods to obtain paginated results. The methods take parameters
+    /// expressing how individual pages are fetched, allowing the same paginator to be used across
+    /// multiple contexts (for example, with different client instances).
+    /// </para>
+    /// </remarks>
+    /// <typeparam name="TResource">The type of resource being paginated</typeparam>
+    /// <typeparam name="TRequest">The type of request used to fetch pages</typeparam>
+    /// <typeparam name="TResponse">The type of response obtained when fetching pages</typeparam>
+    /// <typeparam name="TToken">The type of the "next page token", which must be equatable to itself</typeparam>
+    public sealed class PageStreamer<TResource, TRequest, TResponse, TToken>
+        where TToken : IEquatable<TToken>
+    {
```

Comment:
I saw the example you have at https://github.com/GoogleCloudPlatform/gcloud-dotnet/blob/master/GoogleApiWrappers/src/Google.Apis.Storage.v1.ClientWrapper/StorageClient.ListBuckets.cs#L17 and to be honest it seems complicated, I didn't follow all earlier discussions (sorry!) and I don't have a better suggestion right now, but as someone that sees it for the first time, it seems that we have several funcs, and as a future user of it I would be scared to use it, while I can do the NextToken thing myself.

With that being said, maybe we should re-evaluate trying to include it in the generated code, so the user isn't exposes to this complexity.

Sorry for adding this comment too late in the process... I'm just thinking out loudly :)